packages/flutter/lib/src/cupertino/sheet.dart

```diff
@@ -511,3 +545,208 @@ mixin _CupertinoSheetRouteTransitionMixin<T> on PageRoute<T> {
     return buildPageTransitions<T>(this, context, animation, secondaryAnimation, child);
   }
 }
+
+class _CupertinoDownGestureDetector<T> extends StatefulWidget {
+  const _CupertinoDownGestureDetector({
+    super.key,
+    required this.enabledCallback,
+    required this.onStartPopGesture,
+    required this.child,
+  });
+
+  final Widget child;
+
+  final ValueGetter<bool> enabledCallback;
+
+  final ValueGetter<_CupertinoDownGestureController<T>> onStartPopGesture;
+
+  @override
+  _CupertinoDownGestureDetectorState<T> createState() => _CupertinoDownGestureDetectorState<T>();
+}
+
+class _CupertinoDownGestureDetectorState<T> extends State<_CupertinoDownGestureDetector<T>> {
+  _CupertinoDownGestureController<T>? _downGestureController;
+
+  late VerticalDragGestureRecognizer _recognizer;
+
+  @override
+  void initState() {
+    super.initState();
+    _recognizer =
+        VerticalDragGestureRecognizer(debugOwner: this)
+          ..onStart = _handleDragStart
+          ..onUpdate = _handleDragUpdate
+          ..onEnd = _handleDragEnd
+          ..onCancel = _handleDragCancel;
+  }
+
+  @override
+  void dispose() {
+    _recognizer.dispose();
+
+    // If this is disposed during a drag, call navigator.didStopUserGesture.
+    if (_downGestureController != null) {
+      WidgetsBinding.instance.addPostFrameCallback((_) {
+        if (_downGestureController?.navigator.mounted ?? false) {
+          _downGestureController?.navigator.didStopUserGesture();
+        }
+        _downGestureController = null;
+      });
+    }
+    super.dispose();
+  }
+
+  void _handleDragStart(DragStartDetails details) {
+    assert(mounted);
+    assert(_downGestureController == null);
+    _downGestureController = widget.onStartPopGesture();
+  }
+
+  void _handleDragUpdate(DragUpdateDetails details) {
+    assert(mounted);
+    assert(_downGestureController != null);
+    _downGestureController!.dragUpdate(
+      _convertToLogical(
+        details.primaryDelta! / (context.size!.height - (context.size!.height * 0.08)),
+      ),
+    );
+  }
+
+  void _handleDragEnd(DragEndDetails details) {
+    assert(mounted);
+    assert(_downGestureController != null);
+    _downGestureController!.dragEnd(
+      _convertToLogical(details.velocity.pixelsPerSecond.dy / context.size!.height),
+    );
+    _downGestureController = null;
+  }
+
+  void _handleDragCancel() {
+    assert(mounted);
+    // This can be called even if start is not called, paired with the "down" event
+    // that we don't consider here.
+    _downGestureController?.dragEnd(0.0);
+    _downGestureController = null;
+  }
+
+  void _handlePointerDown(PointerDownEvent event) {
+    if (widget.enabledCallback()) {
+      _recognizer.addPointer(event);
+    }
+  }
+
+  double _convertToLogical(double value) {
```

Comment:
I'm probably missing something obvious, but why does text direction matter for the vertical drag gesture?